storm-server/src/main/java/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java

```diff
@@ -266,4 +286,15 @@ private Object initializeInstance(String className, String representation) {
     public Set<String> getBlacklistSupervisorIds() {
         return Collections.unmodifiableSet(blacklistedSupervisorIds);
     }
+
+    @Override
+    public void nodeAssignmentSent(String node, boolean successful) {
+        if (!blacklistSendAssignentFailures) {
+            return;
+        }
+        if (!successful) {
+            int failCount = assignmentFailures.getOrDefault(node, 0) + 1;
+            assignmentFailures.put(node, failCount);
```

Comment:
Synchronization is needed otherwise there can be a race condition between getting the current count and updating the count (line 296 and line297)

Also between `this.assignmentFailures.clear();` and `assignmentFailures.put(node, failCount);